src/Microsoft.AspNetCore.Sockets.Client/Connection.cs

```diff
@@ -28,58 +29,110 @@ private Connection(Uri url, ITransport transport, IChannelConnection<Message> tr
-        public static Task<Connection> ConnectAsync(Uri url, ITransport transport, HttpClient httpClient) => ConnectAsync(url, transport, httpClient, NullLoggerFactory.Instance);
-
-        public static async Task<Connection> ConnectAsync(Uri url, ITransport transport, HttpClient httpClient, ILoggerFactory loggerFactory)
+        public async Task<bool> ReceiveAsync(ReceiveData receiveData, CancellationToken cancellationToken)
         {
-            if (url == null)
+            if (receiveData == null)
             {
-                throw new ArgumentNullException(nameof(url));
+                throw new ArgumentNullException(nameof(receiveData));
             }
 
-            if (transport == null)
+            if (Input.Completion.IsCompleted)
             {
-                throw new ArgumentNullException(nameof(transport));
+                throw new InvalidOperationException("Cannot receive messages when the connection is stopped.");
             }
 
-            if (httpClient == null)
+            try
             {
-                throw new ArgumentNullException(nameof(httpClient));
+                while (await Input.WaitToReadAsync(cancellationToken))
+                {
+                    if (Input.TryRead(out Message message))
+                    {
+                        using (message)
+                        {
+                            receiveData.Format = message.MessageFormat;
+                            receiveData.Data = message.Payload.Buffer.ToArray();
+                            return true;
+                        }
+                    }
+                }
+
+                await Input.Completion;
             }
+            catch (OperationCanceledException)
+            {
+                // channel is being closed
+            }
+            catch (Exception ex)
+            {
+                Output.TryComplete(ex);
+                _logger.LogError("Error receiving message: {0}", ex);
+                throw;
+            }
+
+            return false;
+        }
 
-            if (loggerFactory == null)
+        public Task SendAsync(byte[] data, Format format)
+        {
+            return SendAsync(data, format, CancellationToken.None);
+        }
+
+        public async Task SendAsync(byte[] data, Format format, CancellationToken cancellationToken)
+        {
+            var message = new Message(ReadableBuffer.Create(data).Preserve(), format);
+
+            while (await Output.WaitToWriteAsync(cancellationToken))
             {
-                throw new ArgumentNullException(nameof(loggerFactory));
+                if (Output.TryWrite(message))
+                {
+                    return;
+                }
             }
 
-            var logger = loggerFactory.CreateLogger<Connection>();
-            var negotiateUrl = Utils.AppendPath(url, "negotiate");
+            throw new InvalidOperationException("Cannot send messages when the connection is stopped.");
```

Comment:
OK. Should we return a flag indicating that the channel was closed (i.e. sending data was not even attempted)?